File: api/src/controllers/userController.js
import User from '../models/UserModel.js';
import jwt from 'jsonwebtoken';

const generateToken = (userId) => {
    return jwt.sign(
        { userId },
        process.env.JWT_SECRET || 'your-secret-key',
        { expiresIn: process.env.JWT_EXPIRES_IN || '7d' }
    );
};

const register = async (req, res) => {
    try {
        const { name, email, password, role } = req.body;

        if (!name || !email || !password) {
            return res.status(400).json({
                success: false,
                message: 'Name, email and password are required'
            });
        }

        if (password.length < 6) {
            return res.status(400).json({
                success: false,
                message: 'Password must be at least 6 characters long'
            });
        }

        const existingUser = await User.findOne({ email });
        if (existingUser) {
            return res.status(400).json({
                success: false,
                message: 'User already exists with this email'
            });
        }

        const user = new User({
            name,
            email,
            password,
            role: role || 'user'
        });

        await user.save();

        const token = generateToken(user._id);

        await user.updateLastLogin();

        res.status(201).json({
            success: true,
            message: 'User registered successfully',
            data: {
                user: {
                    id: user._id,
                    name: user.name,
                    email: user.email,
                    role: user.role,
                    status: user.status,
                    lastLogin: user.lastLogin
                },
                token
            }
        });
    } catch (error) {
        console.error('Register error:', error);
        res.status(500).json({
            success: false,
            message: 'Error registering user',
            error: error.message
        });
    }
};


const login = async (req, res) => {
    try {
        const { email, password } = req.body;

        if (!email || !password) {
            return res.status(400).json({
                success: false,
                message: 'Email and password are required'
            });
        }

        const user = await User.findOne({ email });
        if (!user) {
            return res.status(401).json({
                success: false,
                message: 'Invalid email or password'
            });
        }

        if (user.status !== 'Active') {
            return res.status(401).json({
                success: false,
                message: 'Your account is not active. Please contact administrator.'
            });
        }

        const isPasswordValid = await user.comparePassword(password);
        if (!isPasswordValid) {
            return res.status(401).json({
                success: false,
                message: 'Invalid email or password'
            });
        }

        const token = generateToken(user._id);

        await user.updateLastLogin();

        res.json({
            success: true,
            message: 'Login successful',
            data: {
                user: {
                    id: user._id,
                    name: user.name,
                    email: user.email,
                    role: user.role,
                    status: user.status,
                    lastLogin: user.lastLogin
                },
                token
            }
        });

    } catch (error) {
        console.error('Login error:', error);
        res.status(500).json({
            success: false,
            message: 'Error logging in',
            error: error.message
        });
    }
};


const getProfile = async (req, res) => {
    try {
        res.json({
            success: true,
            data: {
                user: req.user
            }
        });
    } catch (error) {
        console.error('Get profile error:', error);
        res.status(500).json({
            success: false,
            message: 'Error fetching profile',
            error: error.message
        });
    }
};


const getUsers = async (req, res) => {
    try {
        const { page = 1, limit = 10, role, status, search } = req.query;

        const filter = {};

        if (role) filter.role = role;
        if (status) filter.status = status;
        if (search) {
            filter.$or = [
                { name: { $regex: search, $options: 'i' } },
                { email: { $regex: search, $options: 'i' } }
            ];
        }

        const users = await User.find(filter)
            .select('-password')
            .limit(limit * 1)
            .skip((page - 1) * limit)
            .sort({ createdAt: -1 });

        const total = await User.countDocuments(filter);

        res.json({
            success: true,
            data: {
                users,
                totalPages: Math.ceil(total / limit),
                currentPage: parseInt(page),
                totalUsers: total
            }
        });
    } catch (error) {
        console.error('Get users error:', error);
        res.status(500).json({
            success: false,
            message: 'Error fetching users',
            error: error.message
        });
    }
};


const getUser = async (req, res) => {
    try {
        const user = await User.findById(req.params.id).select('-password');

        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        res.json({
            success: true,
            data: { user }
        });
    } catch (error) {
        console.error('Get user error:', error);
        res.status(500).json({
            success: false,
            message: 'Error fetching user',
            error: error.message
        });
    }
};


const updateUser = async (req, res) => {
    try {
        const { name, role, status } = req.body;

        if (req.user._id.toString() === req.params.id && (role || status)) {
            return res.status(403).json({
                success: false,
                message: 'You cannot change your own role or status'
            });
        }

        const user = await User.findById(req.params.id);

        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        if (name) user.name = name;
        if (role) user.role = role;
        if (status) user.status = status;

        await user.save();

        res.json({
            success: true,
            message: 'User updated successfully',
            data: {
                user: {
                    id: user._id,
                    name: user.name,
                    email: user.email,
                    role: user.role,
                    status: user.status,
                    lastLogin: user.lastLogin
                }
            }
        });
    } catch (error) {
        console.error('Update user error:', error);
        res.status(500).json({
            success: false,
            message: 'Error updating user',
            error: error.message
        });
    }
};


const deleteUser = async (req, res) => {
    try {
        if (req.user._id.toString() === req.params.id) {
            return res.status(403).json({
                success: false,
                message: 'You cannot delete your own account'
            });
        }

        const user = await User.findByIdAndDelete(req.params.id);

        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        res.json({
            success: true,
            message: 'User deleted successfully'
        });
    } catch (error) {
        console.error('Delete user error:', error);
        res.status(500).json({
            success: false,
            message: 'Error deleting user',
            error: error.message
        });
    }
};


const logout = async (req, res) => {
    try {
        res.json({
            success: true,
            message: 'Logged out successfully'
        });
    } catch (error) {
        console.error('Logout error:', error);
        res.status(500).json({
            success: false,
            message: 'Error logging out',
            error: error.message
        });
    }
};

export {
    register,
    login,
    logout,
    getProfile,
    getUsers,
    getUser,
    updateUser,
    deleteUser,
};

File: services/app/controllers/user_controller.py
from app.database import db
from app.models.counter_model import get_next_id
from app.utils.hash import hash_password, verify_password
from app.utils.jwt_handler import create_token
from datetime import datetime
import re


class UserController:

    async def register(self, payload):
        name, email, password, role = payload.name, payload.email, payload.password, getattr(payload, "role", "user")

        if not name or not email or not password:
            return {"success": False, "message": "Name, email and password are required"}

        if not re.match(r"^[^\s@]+@[^\s@]+\.[^\s@]+$", email):
            return {"success": False, "message": "Please provide a valid email address"}

        if len(password) < 6:
            return {"success": False, "message": "Password must be at least 6 characters long"}

        existing_user = await db.users.find_one({"email": email})
        if existing_user:
            return {"success": False, "message": "User already exists with this email"}

        user_id = await get_next_id("user")
        now = datetime.utcnow()

        user = {
            "_id": user_id,
            "name": name,
            "email": email,
            "password": hash_password(password),
            "role": role,
            "status": "Active",
            "lastLogin": None,
            "createdAt": now,
            "updatedAt": now
        }

        await db.users.insert_one(user)
        token = create_token(user_id)

        return {
            "success": True,
            "message": "User registered successfully. Please login to continue.",
            "data": {
                "user": {
                    "id": user_id,
                    "name": name,
                    "email": email,
                    "role": role,
                    "status": "Active",
                    "createdAt": now
                },
                "token": token
            }
        }

    async def login(self, payload):
        email, password = payload.email, payload.password

        if not email or not password:
            return {"success": False, "message": "Email and password are required"}

        user = await db.users.find_one({"email": email})

        if not user or not verify_password(password, user["password"]):
            return {"success": False, "message": "Invalid email or password"}

        if user.get("status") != "Active":
            return {"success": False, "message": "Your account is not active. Please contact administrator."}

        token = create_token(user["_id"])
        await db.users.update_one({"_id": user["_id"]}, {"$set": {"lastLogin": datetime.utcnow()}})

        user_data = {k: v for k, v in user.items() if k != "password"}

        return {
            "success": True,
            "message": "Login successful",
            "data": {
                "user": user_data,
                "token": token
            }
        }

    async def get_current_user(self, user_id: int):
        user = await db.users.find_one({"_id": user_id})
        if not user:
            return {"success": False, "message": "User not found"}
        user_data = {k: v for k, v in user.items() if k != "password"}
        return {"success": True, "data": {"user": user_data}}

    async def get_users(self, skip: int = 0, limit: int = 100):
        cursor = db.users.find().skip(skip).limit(limit)
        users = []
        async for user in cursor:
            user_data = {k: v for k, v in user.items() if k != "password"}
            users.append(user_data)
        return {"success": True, "data": {"users": users}}

    async def get_user(self, user_id: str):
        user = await db.users.find_one({"_id": user_id})
        if not user:
            return {"success": False, "message": "User not found"}
        user_data = {k: v for k, v in user.items() if k != "password"}
        return {"success": True, "data": {"user": user_data}}

    async def update_user(self, user_id: str, payload: dict):
        if "password" in payload:
            payload["password"] = hash_password(payload["password"])

        payload["updatedAt"] = datetime.utcnow()

        result = await db.users.update_one({"_id": user_id}, {"$set": payload})
        if result.matched_count == 0:
            return {"success": False, "message": "User not found"}

        updated_user = await db.users.find_one({"_id": user_id})
        user_data = {k: v for k, v in updated_user.items() if k != "password"}
        return {"success": True, "message": "User updated successfully", "data": {"user": user_data}}

    async def delete_user(self, user_id: str):
        result = await db.users.delete_one({"_id": user_id})
        if result.deleted_count == 0:
            return {"success": False, "message": "User not found"}
        return {"success": True, "message": "User deleted successfully"}
